File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* removeNthFromEnd(struct ListNode* head, int n) {
    struct ListNode* dummy = malloc(sizeof(struct ListNode));
    dummy->next = head;

    struct ListNode* fast = dummy;
    struct ListNode* slow = dummy;

    // Move fast k steps
    for (int i = 1; i <= n; i++)
    {
        fast = fast->next;
    }

    // Move both until fast.next reaches NULL
    while (fast->next!= NULL)
    {
        fast = fast->next;
        slow = slow->next;
    }

    // slow->next is the node to delete
    struct ListNode* temp = slow->next;

    slow->next = temp->next;

    free(temp);

    head = dummy->next;
    free(dummy);

    return head;
}

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode fast = head;
        ListNode slow = head;
        ListNode temp;
        for(int i=1;i<=n;i++){
            fast=fast.next;
        }

        if(fast==null){
            return head.next;
        }
        while(fast.next!=null){
            slow=slow.next;
            fast=fast.next;
        }
        temp=slow.next;
        slow.next=temp.next;
        return head;
    }
}